### PR TITLE
Add RbfProfileParameter.

### PR DESCRIPTION
**Pull request by jetuk**
See docstring of new parameter for information.

### CODE REVIEW

**Comment by jetuk** [comment]
Note that this is a more specific form of what's in #725 .

File: pywr/parameters/_parameters.pyx [hunk 1]
```diff
@@ -905,6 +906,109 @@ cdef class UniformDrawdownProfileParameter(Parameter):
...
+        self._interpolate()
+
+    cpdef _interpolate(self):
+        cdef int i
+
+        days_of_year = list(self._days_of_year)
+        # Append day 365 to the list and mirror the penultimate and second DOY at the start and end
+        # of the list respectively. This helps ensure the gradient is roughly the same across the boundary
+        # between days 365 and 0.
+        days_of_year = [days_of_year[-1]-365+1] + list(days_of_year) + [365, 365+days_of_year[1]-1]
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx:965]: If one of the input 'days_of_year' is 365 then adding another one here will cause an error. probably needs a check/warning in the docstring/init

File: pywr/parameters/_parameters.pyx [hunk 2]
```diff
@@ -905,6 +906,109 @@ cdef class UniformDrawdownProfileParameter(Parameter):
...
+    In addition the second and penultimate values are mirrored to encourage a consistent
+    gradient to appear across the boundary. The RBF calculations are undertaken using
+    the `scipy.interpolate.Rbf` object, please refer to Scipy's documentation for more
+    information.
+
+    Parameters
+    ----------
+    days_of_year : iterable, integer
+        The days of the year at which the interpolation values are defined. The first
+        value should be one.
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx:925]: Worth raising an exception if the first value is not 1?

File: pywr/parameters/_parameters.pyx [hunk 3]
```diff
@@ -905,6 +906,109 @@ cdef class UniformDrawdownProfileParameter(Parameter):
...
+    """
+    def __init__(self, model, days_of_year, values, lower_bounds=0.0, upper_bounds=np.inf, rbf_kwargs=None, **kwargs):
+        super(RbfProfileParameter, self).__init__(model, **kwargs)
+
+        if len(days_of_year) != len(values):
+            raise ValueError(f"The length of values ({len(values)}) must equal the length of "
+                             f"`days_of_year` ({len(days_of_year)}).")
+
+        self.double_size = len(values)
+        self.integer_size = 0
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx:944]: Already set to 0 in `Parameter`?

File: pywr/parameters/_parameters.pyx (hunk 1 above)
Comment (jetuk) [reply pywr/parameters/_parameters.pyx:965]: I think this is incorrect actually. It should be repeating day 1 as value 366 - I'm assuming zero based indexing here.

File: pywr/parameters/_parameters.pyx (hunk 2 above)
Comment (jetuk) [reply pywr/parameters/_parameters.pyx:925]: Good idea.

File: pywr/parameters/_parameters.pyx (hunk 3 above)
Comment (jetuk) [reply pywr/parameters/_parameters.pyx:944]: Just being explicit :)

File: pywr/parameters/_parameters.pyx (hunk 1 above)
Comment (Batch21) [reply pywr/parameters/_parameters.pyx:965]: yeah ok, forgot about the zero indexing!

Comment (jetuk) [reply pywr/parameters/_parameters.pyx]: This should be fixed now. I've added some checking of the days of the year.

File: pywr/parameters/_parameters.pyx (hunk 2 above)
Comment (jetuk) [reply pywr/parameters/_parameters.pyx:925]: Done.

Comment (jetuk) [reply pywr/parameters/_parameters.pyx:925]: Actually, I guess this could be implemented without this being a requirement?

**Comment by jetuk** [comment]
@Batch21 I've updated this. The interpolation 1-based indexing should make sense, and there are input checks on the `days_of_year`. However, I just thought whether the requirement for the first day to be 1 is really needed?

**Comment by Batch21** [comment]
because it assumes the users wants the boundary to be as smooth as possible? Could just be advice in the docstring rather than a hard constraint

**Comment by jetuk** [comment]
I think you could retain the mirroring (for the boundary effect) without requiring the first day to be 1.

**Comment by jetuk** [comment]
It does make it much more controlled at the boundary, so perhaps a good idea to keep it as is.

**Comment by Batch21** [comment]
I'm happy for it to be kept. Had a quick play around, and even with the mirroring you don't have to move the index of the first value that far away from 1 to get quite a jump at the boundary

**Comment by jetuk** [comment]
OK. So shall we merge this as is?

**Pull request merged**